### PR TITLE
[`Mistral`] Add Flash Attention-2 support for `mistral`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Adds Flash Attention 2 for Mistral For Causal - we still need to discuss how to integrate it with local attention

cc @ArthurZucker @LysandreJik 

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1",
    use_flash_attention_2=True,
    torch_dtype=torch.float16,
    low_cpu_mem_usage=True
).to(0)

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")

text = "Hello my name is"
inputs = tokenizer(text, return_tensors="pt").to(0)

out = model.generate(**inputs, max_new_tokens=4096, use_cache=True, do_sample=True)
print(tokenizer.batch_decode(out, skip_special_tokens=True))
```

### CODE REVIEW

**Comment by timlacroix** [comment]
Might be worth adding support for sliding window attention ? 
params are window_size_left = config.sliding_window-1 // window_size_right = -1 I believe ? 
See this PR adding support in flashattentionv2
https://github.com/Dao-AILab/flash-attention/commit/083e8f525f1f8e1dde5044afdac79f9588302207

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -291,11 +319,208 @@ def forward(
...
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value[0].shape[-2]
+        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+
+        use_sliding_windows = _is_flash_using_slicing_windows and kv_seq_len > self.config.sliding_window and not self.training
```
Comment (timlacroix) [inline src/transformers/models/mistral/modeling_mistral.py:356]: typo, should be "_is_flash_using_sliding_windows"

Comment (timlacroix) [reply src/transformers/models/mistral/modeling_mistral.py:356]: also I don't think this bool is needed. 

If self.config.sliding_window is not None -> use sliding_window always, whether training or inferencing no ?

Comment (younesbelkada) [reply src/transformers/models/mistral/modeling_mistral.py:356]: Hmm good point, for some reason I thought that feature works only for inference (from the source code's readme: https://github.com/mistralai/mistral-src#sliding-window-to-speed-up-inference-and-reduce-memory-pressure I have read "speed up inference" so I thought that was only available for inference) - will remove that condition

File: src/transformers/models/mistral/modeling_mistral.py [hunk 2]
```diff
@@ -291,11 +319,208 @@ def forward(
...
+                    value_states,
+                    cu_seqlens_q=cu_seqlens_q,
+                    cu_seqlens_k=cu_seqlens_k,
+                    max_seqlen_q=max_seqlen_in_batch_q,
+                    max_seqlen_k=max_seqlen_in_batch_k,
+                    dropout_p=dropout,
+                    softmax_scale=softmax_scale,
+                    causal=True,
+                    window_size=(self.config.sliding_window, self.config.sliding_window)
+                )
```
Comment (timlacroix) [inline src/transformers/models/mistral/modeling_mistral.py:464]: could probably be factored ? something like window_size=(self.config.sliding_window or -1, -1) ?

File: src/transformers/models/mistral/modeling_mistral.py [hunk 3]
```diff
@@ -291,11 +319,208 @@ def forward(
...
+
+            attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
+        else:
+            if not use_sliding_windows:
+                attn_output = flash_attn_func(
+                    query_states, key_states, value_states, dropout, softmax_scale=softmax_scale, causal=True
+                )
+            else:
+                attn_output = flash_attn_func(
+                    query_states, key_states, value_states, dropout, softmax_scale=softmax_scale, causal=True, window_size=(self.config.sliding_window // 2, self.config.sliding_window // 2)
```
Comment (timlacroix) [inline src/transformers/models/mistral/modeling_mistral.py:474]: same comment on factoring.
I have no idea what's going on with // 2 here but I don't know what padding_mask is :|

Comment (younesbelkada) [reply src/transformers/models/mistral/modeling_mistral.py:474]: Oh yeah ignore the //2 I used it for testing purpose :D

Comment (younesbelkada) [reply src/transformers/models/mistral/modeling_mistral.py:474]: padding mask is the "pure" attention mask (not causal mask) 0 if padding token 1 if not --> I use it in the control flow for flash attention modules whether I need to pad / unpad or no

File: src/transformers/models/mistral/modeling_mistral.py [hunk 4]
```diff
@@ -291,11 +321,219 @@ def forward(
...
+
+        if past_key_value is not None:
+            if use_sliding_windows and kv_seq_len > self.config.sliding_window:
+                slicing_tokens = kv_seq_len - self.config.sliding_window
+
+                past_key = past_key_value[0]
+                past_value = past_key_value[1]
+
+                past_key = past_key[:, :, :-slicing_tokens, :].contiguous()            
+                past_value = past_value[:, :, :-slicing_tokens, :].contiguous()
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:370]:
```suggestion
                past_key = past_key[:, :, 1:, :].contiguous()            
                past_value = past_value[:, :, 1:, :].contiguous()  
```
this makes more sense no?

**Comment by younesbelkada** [comment]
Sharing some results here vs mistral official implementation that uses `xformers.memory_efficient_attention`: 

##  Context length = 12 / max_new_tokens=512 / bs=1

HF transformers + FA-2
```
Latency: 15.1241201171875
33.85320904838279 tokens / s
Max allocated memory: 15218032640
```
Mistral + mem efficient:
```
Latency: 17.23331640625
29.709893785407036 tokens / s
Max allocated memory: 14636799488
```

## Context length = 11K / max_new_tokens=512 / bs=1

HF transformers + FA-2
```
Latency: 16.497216796875
31.03553807312431 tokens / s
Max allocated memory: 18673463808 
```
Mistral + mem efficient:
```
Latency: 22.50997265625
22.74547409802565 tokens / s
Max allocated memory: 17303250944
```

## Context length = 11K / max_new_tokens=512 / bs=2 with 11K padding tokens on the second batch

HF transformers + FA-2
```
Latency: 33.95778515625
15.077544004832287 tokens / s
Max allocated memory: 22320273408
```
Mistral + mem efficient:
```
Latency: 30.407841796875
16.83776189774238 tokens / s
Max allocated memory: 17841224192
```

## Context length = 11K / max_new_tokens=512 / bs=4 with 11K padding tokens on the second, third and fourth batch

HF transformers + FA-2
```
Latency: 48.86058984375
10.478792860203109 tokens / s
Max allocated memory: 29610738688
```
Mistral + mem efficient:
```
Latency: 45.27477734375
11.308724858272097 tokens / s
Max allocated memory: 18914968576
```

--> obviously the pad / unpad overhead takes it over for the HF implementation whereas the official repository deals with padding tokens differently. Note also that the max allocated memory increases if one adds padding token. Also note the current cache slicing mechanism assumes users are under `padding=left` regime. Generation should be performed with padding_side=left whereas this should have no impact for training as the cache is not used during training.

Here is a plot that compares pure forward on HF native vs HF + FA-2

Screenshot 2023-10-02 at 17 49 12


Screenshot 2023-10-02 at 17 49 06

**Comment by younesbelkada** [comment]
Script I used to benchmark transformers + FA2: https://gist.github.com/younesbelkada/691c1dec3da2f0a7de29c1d1096d860f 

Script I used to benchmark mistral original source code: https://gist.github.com/younesbelkada/ada0d9c2c48ab034486dbaaf95d29fae (assumes you have cloned their repository and run it under the root folder of the repo)

File: docs/source/en/perf_infer_gpu_one.md [hunk 5]
```diff
@@ -32,6 +32,7 @@ Make sure to follow the installation guide on the repository mentioned above to
 We natively support Flash Attention 2 for the following models:
 
 - Llama
+- Mistral
```
Comment (ArthurZucker) [inline docs/source/en/perf_infer_gpu_one.md:35]: As I mentioned before, I think it is important to add the expected improvement from this addition. Let's add `MistralFlashAttention` to the `mistral.md` with a small section on expected gains and observed gains etc with the result and benchmark you ran. Specifying that it depends on the hardware etc etc

File: src/transformers/models/mistral/modeling_mistral.py [hunk 6]
```diff
@@ -18,26 +18,53 @@
...
+    replace_return_docstrings,
+)
 from .configuration_mistral import MistralConfig
 
 
+if is_flash_attn_available():
+    from flash_attn import flash_attn_func, flash_attn_varlen_func
+    from flash_attn.bert_padding import index_first_axis, pad_input, unpad_input  # noqa
+
+    _is_flash_using_sliding_windows = "window_size" in list(inspect.signature(flash_attn_func).parameters)
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:48]: Might be better to check the version of the package wdyt? Should we require it as well? IMO yes 
```suggestion
    _flash_supports_window_size = "window_size" in list(inspect.signature(flash_attn_func).parameters)
```

File: src/transformers/models/mistral/modeling_mistral.py [hunk 7]
```diff
@@ -18,26 +18,53 @@
...
+
+
 logger = logging.get_logger(__name__)
 
 _CONFIG_FOR_DOC = "MistralConfig"
 
 
+def _get_unpad_data(padding_mask):
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:56]: missing copied from

File: src/transformers/models/mistral/modeling_mistral.py [hunk 8]
```diff
@@ -291,11 +319,273 @@ def forward(
...
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value[0].shape[-2]
+
+        # Extrapolate the RoPE in case of sliding windows
+        rotary_seq_len = max(kv_seq_len, position_ids[:, -1].max().item()) + 1
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:354]:
```suggestion
        # Because the input can be padded, the absolute sequence length depends on the max position id.
        rotary_seq_len = max(kv_seq_len, position_ids[:, -1].max().item()) + 1
```

File: src/transformers/models/mistral/modeling_mistral.py [hunk 9]
```diff
@@ -291,11 +319,273 @@ def forward(
...
+            _is_flash_using_sliding_windows
+            and hasattr(self.config, "sliding_window") is not None
+            and kv_seq_len > self.config.sliding_window
+        )
+
+        if not _is_flash_using_sliding_windows:
+            logger.warning_once(
+                "The current flash attention version does not support sliding window attention, for a more memory efficient implementation"
+                " make sure to upgrade flash-attn library."
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:369]: should go in the import instead of here

File: src/transformers/models/mistral/modeling_mistral.py [hunk 10]
```diff
@@ -291,11 +319,273 @@ def forward(
...
+        if input_dtype == torch.float32:
+            logger.warning_once(
+                "The input hidden states seems to be silently casted in float32, this might be related to"
+                " the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
+                " float16."
+            )
+
+            query_states = query_states.to(torch.float16)
+            key_states = key_states.to(torch.float16)
+            value_states = value_states.to(torch.float16)
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:423]: not sure if this should be included since its peft only always casts to float16 even if input is bfloat16

File: src/transformers/models/mistral/modeling_mistral.py [hunk 11]
```diff
@@ -569,11 +862,26 @@ def forward(
...
+            padding_mask = attention_mask
+
+        if padding_mask is not None and hasattr(self.config, "_flash_attn_2_enabled"):
+            is_padding_right = padding_mask[:, -1].sum().item() != batch_size
+            if not is_padding_right:
+                logger.warning_once(
+                    "You are attempting to perform batched generation with padding_side='right'"
+                    " this may lead to unexpected behaviour for Flash Attention version of Mistral. Make sure to "
+                    " call `tokenizer.padding_side  = 'left'` before tokenizing the input. "
+                )
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:883]: given how bad the outputs were might be a good idea to for padding side / raise and error

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot! Looking good.

File: src/transformers/models/mistral/modeling_mistral.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/models/mistral/modeling_mistral.py:424]: `MistralRMSNorm` behaves exactly as LLamaRMSNorm so it will silently cast the hidden states in fp32, therefore this is needed. As mentioned offline I will address a proper fix for bf16 issues

**Review by ArthurZucker — APPROVED** [review]
Thanks of iterating!

File: src/transformers/models/mistral/modeling_mistral.py [hunk 12]
```diff
@@ -291,11 +320,271 @@ def forward(
...
+        if not _flash_supports_window_size:
+            logger.warning_once(
+                "The current flash attention version does not support sliding window attention, for a more memory efficient implementation"
+                " make sure to upgrade flash-attn library."
+            )
+
+        if past_key_value is not None:
+            # Activate slicing cache only if the config has a value `sliding_windows` attribute
+            if hasattr(self.config, "sliding_window") and kv_seq_len > self.config.sliding_window:
+                slicing_tokens = kv_seq_len - self.config.sliding_window
```
Comment (vince62s) [inline src/transformers/models/mistral/modeling_mistral.py:375]: @younesbelkada sorry to bother again but, if kv_seq_len > N times sliding_windows for instance seq_len = 9000 and sliding_window = 4096 shouldn't slicing_tokens be 9000 - 2 x 4096 instead of 9000 - 4096 ?

Comment (younesbelkada) [reply src/transformers/models/mistral/modeling_mistral.py:375]: Hmm possibly yes, I need to double check with the original code, I'll get back to you on this!

File: src/transformers/models/mistral/modeling_mistral.py (hunk 11 above)
Comment (dakinggg) [reply src/transformers/models/mistral/modeling_mistral.py]: @younesbelkada the implementation here doesn't seem specific to generation in any way. Is the error message wrong or is the implementation wrong (or am I missing something)? That is, should I be able to run forward with right padding and flash2?

File: src/transformers/models/mistral/modeling_mistral.py (hunk 12 above)
Comment (younesbelkada) [reply src/transformers/models/mistral/modeling_mistral.py:375]: @vince62s I could not find any relevant piece of code in the source code of mistral to confirm your statement, can you help me identifying the place where you think we indeed need to slice only N*4096?

Comment (vince62s) [reply src/transformers/models/mistral/modeling_mistral.py:375]: I think I was mistaken. We are adding key/value one by one here so kv_seq_len is never > self.config.sliding_window + 1 it is exactly equal to. I was misled my line 374.
